docs/plugin-authoring-evaluator.md

```diff
@@ -0,0 +1,380 @@
+
+This example leverages `handlebars` to hydrate the prompt and `zod` to format and validate the response.
+
+```ts
+// Define the expected output values
+const DELICIOUSNESS_VALUES = ['yes', 'no', 'maybe'] as const;
+
+// Define the response schema expected from the LLM
+const DeliciousnessDetectionResponseSchema = z.object({
+  reason: z.string(),
+  verdict: z.enum(DELICIOUSNESS_VALUES),
+});
+
+type DeliciousnessDetectionResponse = z.infer<
+  typeof DeliciousnessDetectionResponseSchema
+>;
+
+/**
+ * Score an individual test case for delciousness.
+ */
+export async function deliciousnessScore<
+  CustomModelOptions extends z.ZodTypeAny,
+>(
+  judgeLlm: ModelArgument<CustomModelOptions>,
+  dataPoint: BaseDataPoint,
+  judgeConfig?: CustomModelOptions
+): Promise<Score> {
+  const d = dataPoint;
+  try {
+    // Validate the input has required fields
+    if (!d.output) {
+      throw new Error('Output is required for Funniness detection');
+    }
+
+    //Hydrate the prompt
+    const promptTemplate = Handlebars.compile(QUESTION_GEN_PROMPT);
+    const finalPrompt = promptTemplate({
+      output: d.output,
+    });
+
+    // Call the LLM to generate an evaluation result
+    const response = await generate({
+      model: judgeLlm,
+      prompt: finalPrompt,
+      config: judgeConfig,
+    });
+
+    // Parse the output
+    const parsedResponse = extractJson<DeliciousnessDetectionResponse>(
+      response.text()
+    );
+    if (!parsedResponse) {
+      throw new Error(`Unable to parse evaluator response: ${response.text()}`);
+    }
+
+    // Return a scored response
+    return {
+      score: parsedResponse.verdict,
+      details: { reasoning: parsedResponse.reason },
+    };
+  } catch (err) {
+    console.debug(
+      `Deliciousness evaluation failed with error ${err} for sample ${JSON.stringify(
+        d
+      )}`
+    );
+    throw err;
+  }
+}
+```
+
+#### Define the evaluator action
+
+The final step is to write a function that defines the evaluator action itself.
+
+```ts
+/**
+ * Create the Deliciousness evaluator action.
+ */
+export function createDeliciousnessEvaluator<
+  ModelCustomOptions extends z.ZodTypeAny,
+>(
+  judge: ModelReference<ModelCustomOptions>,
+  judgeConfig: z.infer<ModelCustomOptions>
+): EvaluatorAction {
+  return defineEvaluator(
+    {
+      name: `myAwesomeEval/deliciousness`,
+      displayName: 'Deliciousness',
+      definition: 'Determines if output is considered delicous.',
+    },
+    async (datapoint: BaseDataPoint) => {
+      const score = await deliciousnessScore(judge, datapoint, judgeConfig);
+      return {
+        testCaseId: datapoint.testCaseId,
```

Comment:
Thought: Can we avoid requiring the plugin writer to have to write this boilerplate `testCaseId: datapoint.testCaseId`?

Reply:
I don't know if we can at the moment because it's the way that we map all the additional metadata from the input dataset into the final scored result.